book/visual-effects.md

```diff
@@ -1558,56 +1591,83 @@ the page, so that clipping only applies to that element.
-These control whether `saveLayer` is called and whether subcommands
-are actually painted:
+and we could skip creating those surfaces most of the time. For
+example, there's no reason to create a surface in `Alpha` if no
+opacity is actually applied:
 
 ``` {.python}
-class SaveLayer:
-    def __init__(self, sk_paint, children,
-            should_save=True, should_paint_cmds=True):
-        self.should_save = should_save
-        self.should_paint_cmds = should_paint_cmds
+class Alpha:
+    def execute(self, canvas):
+        paint = skia.Paint(Alphaf=self.alpha)
+        if self.alpha < 1:
+            canvas.saveLayer(paint=paint)
+        for cmd in self.children:
+            cmd.execute(canvas)
+        if self.alpha < 1:
+            canvas.restore()
+```
+
+Similarly, `Blend` doesn't necessarily need to create a layer if
+there's no blending going on. But the logic here is a little trickier.
+Recall `paint_visual_effects`:
+
+``` {.python expected=False}
+def paint_visual_effects(node, cmds, rect):
+    # ...
+   return [
+        Blend(blend_mode, [
+            Alpha(opacity, cmds)
+            Blend("destination-in", [
+                DrawRRect(rect, clip_radius, "white")
+            ]),
+        ]),
+    ]
+```
+
+Here, the outer `Blend` operation not only applies blending but also
+isolates the element contents `cmds` so only they are clipped by
+`overflow`. So let's add an `isolate` parameter to `Blend`, and skip
+blending only if we don't need isolation _and_ we aren't applying
+blending:
+
+``` {.python}
+class Blend:
+    def __init__(self, blend_mode, isolate, children):
+        self.blend_mode = blend_mode
+        self.isolate = isolate
+        self.should_save = self.isolate or self.blend_mode
         # ...
 
     def execute(self, canvas):
+        paint = skia.Paint(
+            BlendMode=parse_blend_mode(self.blend_mode))
         if self.should_save:
-            canvas.saveLayer(paint=self.sk_paint)
-        if self.should_paint_cmds:
-            for cmd in self.children:
-                cmd.execute(canvas)
+            canvas.saveLayer(paint=paint)
+        for cmd in self.children:
+            cmd.execute(canvas)
         if self.should_save:
             canvas.restore()
 ```
 
-Turn off those parameters if an effect isn't applied:
+The `isolate` parameter will indicate whether we're clipping:
 
 ``` {.python expected=False}
 def paint_visual_effects(node, cmds, rect):
     # ...
-
-    needs_clip = node.style.get("overflow", "visible") == "clip"
-    needs_blend_isolation = blend_mode != skia.BlendMode.kSrcOver or \
-        needs_clip
-    needs_opacity = opacity != 1.0
-
    return [
-        SaveLayer(skia.Paint(BlendMode=blend_mode), [
-            SaveLayer(skia.Paint(Alphaf=opacity), cmds,
-                should_save=needs_opacity),
-            SaveLayer(skia.Paint(BlendMode=skia.kDstIn), [
+        Blend(blend_mode, clip_radius > 0, [
```

Comment:
We could avoid the isolation parameter by this; WDYT?

if not blend_mode and clip_radius > 0:
  blend_mode = "source-over"
Blend(blend-mode, ...)

Reply:
Oh yeah, we can! I initially had the parse elsewhere and changed it for unrelated reasons but now this works great!